jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -582,22 +583,29 @@ else if (cause instanceof UnavailableException)
             }
             catch (Throwable e)
             {
-                abort(e);
+                failure.addSuppressed(e);
+                abort(failure);
                 if (LOG.isDebugEnabled())
-                    LOG.debug("Could not commit response error 500", e);
+                    LOG.debug("Could not commit response error 500", failure);
             }
         }
     }
 
-    protected Throwable unwrap(Throwable th)
+    /** Unwrap failure causes to find target class
+     * @param failure The throwable to have its causes unwrapped
+     * @param targets Exception classes that we should not unwrap
+     * @return A target throwable or null
+     */
+    protected Throwable unwrap(Throwable failure, Class<?> ... targets)
     {
-        // Unwrap to root cause:
-        // RuntimeIOExceptions are unwrapped as they are holders of IOExceptions that may be verbosity controlled
-        // ServletExceptions are unwrapped as are just general holders of exceptions
-        // UnavailableException are not unwrapped as they are specialized ServletExceptions that carry extra information.
-        while (th.getCause()!=null && (th instanceof RuntimeIOException || ((th instanceof ServletException) && !(th instanceof UnavailableException))))
-            th = th.getCause();
-        return th;
+        while (failure!=null)
+        {
+            for (Class<?> x : targets)
+                if (x.isInstance(failure))
```

Comment:
If someone extends from BadMessageException or UnavailableException does that change the meaning of it in your mind?

Reply:
Not for the purposes of this unwrap as but BME and UAE provide methods that influence the response code to be sent and any extended classes need  to respect those.